File: BE/Node_Typescript/src/repositories/data/models/Employee.ts
import { Table, Column, Model, DataType, BeforeSave, BelongsTo, BelongsToMany, ForeignKey } from 'sequelize-typescript';

import { Field } from './Field';
import { User } from './User';
import { RecruitmentArticle } from './RecruitmentArticle';
import { Enjoy } from './Enjoy';
import { DetailRecruitment } from './DetailRecruitment';

@Table({ timestamps: false })
class Employee extends Model {
    @Column({ primaryKey: true, type: DataType.STRING })
    @ForeignKey(() => User)
    Id!: string;

    @Column({type: DataType.STRING})
    FullName!: string;

    @Column({type: DataType.STRING})
    PhoneNumber!: string;

    @Column({type: DataType.STRING})
    Avatar!: string;

    @Column({type: DataType.STRING})
    Introduction!: string;

    @Column({ allowNull: true,type: DataType.STRING })
    Certification!: string;

    @Column({ allowNull: true,type: DataType.STRING })
    CV!: string;

    @Column({type: DataType.DATE})
    Born!: Date;

    @Column({type: DataType.STRING})
    Gender!: string;

    @Column({type: DataType.STRING})
    Address!: string;

    @BelongsTo(() => Field, { foreignKey: 'FieldId' })
    Field!: Field;

    @BelongsTo(() => User, { foreignKey: 'Id' })
    User!: User
    //
    @BeforeSave
    static async processCertification(instance: Employee) {
        if (instance.Certification && Array.isArray(instance.Certification)) {
            instance.Certification = instance.Certification.join(',');
        }
    }

    getCertificationArray(): string[] | null {
        if (this.Certification) {
            return this.Certification.split(',');
        }
        return null;
    }

    @BelongsToMany(() => RecruitmentArticle, () => Enjoy)
    ListEnjoy!: RecruitmentArticle[];

    @BelongsToMany(() => RecruitmentArticle, () => DetailRecruitment)
    ListDR!: RecruitmentArticle[];
}

export { Employee };

File: FE/web_admin/lib/models/employee.dart
class Employee {
  final String id;
  final String avatar;
  final String fullName;
  final String introduction;
  final String born;
  final String gender;
  final String address;
  final String phoneNumber;
  final String certification;
  final String field_id;

  @override
  String toString() {
    return '\nEmployee {id: $id, avatar: $avatar, fullName: $fullName, introduction: $introduction, born: $born, gender: $gender, address: $address, phoneNumber: $phoneNumber, certification: $certification}';
  }

  List<String> toArray() => [
        id,
        avatar,
        fullName,
        introduction,
        born,
        gender,
        address,
        phoneNumber,
        certification,
        field_id
      ];

  static List<String> toName() {
    return [
      'Id',
      'Avatar',
      'Full Name',
      'Introduction',
      'Born',
      'Gender',
      'Address',
      'Phone number',
      'Certification',
      'Field ID'
    ];
  }

  Employee(
      {required this.id,
      required this.avatar,
      required this.fullName,
      required this.introduction,
      required this.born,
      required this.gender,
      required this.address,
      required this.phoneNumber,
      required this.certification,
      required this.field_id});

  factory Employee.fromJson(Map<String, dynamic> json) {
    return Employee(
      id: json['id'].toString(),
      avatar: json['avatar'].toString(),
      fullName: json['fullname'].toString(),
      introduction: json['introduction'].toString(),
      born: json['born'].toString(),
      gender: json['gender'].toString(),
      address: json['address'].toString(),
      phoneNumber: json['phoneNumber'].toString(),
      certification: json['certification'].toString(),
      field_id: json['field_id'].toString(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'avatar': avatar,
        'fullName': fullName,
        'introduction': introduction,
        'born': born,
        'gender': gender,
        'address': address,
        'phoneNumber': phoneNumber,
        'certification': certification,
        'field_id': field_id
      };
}
